readthedocs/aws/security_token_service.py

```diff
@@ -0,0 +1,233 @@
+"""
+Module to interact with AWS STS (Security Token Service) to assume a role and get temporary scoped credentials.
+
+This is mainly used to generate temporary credentials to interact with S3 buckets from the builders.
+
+In order to make use of STS, we need:
+
+- Create a role in IAM with a trusted entity type set to the AWS account that is going to be used to generate the temporary credentials.
+- A policy that allows access to all S3 buckets and paths that are going to be used.
+  Which should be attached to the role.
+- The permissions of the temporary credentials are the result of the intersection of the role policy and the inline policy that is passed to the AssumeRole API.
+  This means that the inline policy can be used to limit the permissions of the temporary credentials, but not to expand them.
+
+See:
+
+- https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html
+- https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_sts-comparison.html
+- https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_assumerole.html
+- https://docs.readthedocs.com/dev/latest/aws-temporary-credentials.html
+"""
+
+import json
+from dataclasses import dataclass
+
+import boto3
+import structlog
+from django.conf import settings
+
+
+log = structlog.get_logger(__name__)
+
+
+class AWSTemporaryCredentialsError(Exception):
+    """Exception raised when there is an error getting AWS S3 credentials."""
+
+
+@dataclass
+class AWSTemporaryCredentials:
+    """Dataclass to hold AWS temporary credentials."""
+
+    access_key_id: str
+    secret_access_key: str
+    session_token: str | None
+
+
+@dataclass
+class AWSS3TemporaryCredentials(AWSTemporaryCredentials):
+    """Subclass of AWSTemporaryCredentials to include S3 specific fields."""
+
+    bucket_name: str
+    region_name: str
+
+
+def get_sts_client():
+    return boto3.client(
+        "sts",
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+        # TODO: should this be its own setting?
+        region_name=settings.AWS_S3_REGION_NAME,
```

Comment:
Is this TODO still valid? It seems it's its own setting already.

Reply:
This setting is linked to storage, not the region name of the account used to interact with STS, not sure if it matters anyway...

Reply:
I could try testing without region and see :man_shrugging:

Reply:
We could probably have a AWS_REGION_NAME setting, and AWS_S3_REGION_NAME defaults to that one, but could also just use AWS_S3_REGION_NAME everywhere.